tools/apps/src/lib.rs

```diff
@@ -0,0 +1,900 @@
+
+            _ => panic!("Unsupported HTTP version"),
+        };
+
+        let max_data = args.get_str("--max-data");
+        let max_data = u64::from_str_radix(max_data, 10).unwrap();
+
+        let max_stream_data = args.get_str("--max-stream-data");
+        let max_stream_data = u64::from_str_radix(max_stream_data, 10).unwrap();
+
+        let max_streams_bidi = args.get_str("--max-streams-bidi");
+        let max_streams_bidi = u64::from_str_radix(max_streams_bidi, 10).unwrap();
+
+        let max_streams_uni = args.get_str("--max-streams-uni");
+        let max_streams_uni = u64::from_str_radix(max_streams_uni, 10).unwrap();
+
+        let dump_packet_path = if args.get_str("--dump-packets") != "" {
+            Some(args.get_str("--dump-packets").to_string())
+        } else {
+            None
+        };
+
+        let no_grease = args.get_bool("--no-grease");
+
+        CommonArgs {
+            alpns,
+            max_data,
+            max_stream_data,
+            max_streams_bidi,
+            max_streams_uni,
+            dump_packet_path,
+            no_grease,
+        }
+    }
+}
+
+pub struct PartialResponse {
+    pub body: Vec<u8>,
+
+    pub written: usize,
+}
+
+pub struct Client {
+    pub conn: std::pin::Pin<Box<quiche::Connection>>,
+
+    pub http_conn: Option<Box<dyn crate::HttpConn>>,
+
+    pub partial_responses: std::collections::HashMap<u64, PartialResponse>,
+}
+
+pub type ClientMap = HashMap<Vec<u8>, (net::SocketAddr, Client)>;
+
+// Makes a buffered writer for a resource with a target URL. The
+// file will have the same name as the resource's last path segment value.
+// Multiple requests for the same URL are indicated by the value of `cardinal`,
+// any value "N" greater than 1, will cause ".N" to be appended to the filename.
+fn make_writer(
+    url: &url::Url, target_path: &Option<String>, cardinal: u64,
+) -> Option<std::io::BufWriter<std::fs::File>> {
+    if let Some(tp) = target_path {
+        let resource =
+            url.path_segments().map(|c| c.collect::<Vec<_>>()).unwrap();
+
+        let mut path = format!("{}/{}", tp, resource.iter().last().unwrap());
+
+        if cardinal > 1 {
+            path = format!("{}.{}", path, cardinal);
+        }
+
+        match std::fs::File::create(&path) {
+            Ok(f) => return Some(std::io::BufWriter::new(f)),
+
+            Err(e) => panic!("Bad times: {}", e),
+        }
+    }
+
+    None
+}
+
+pub trait HttpConn {
+    fn send_requests(
+        &mut self, conn: &mut quiche::Connection, target_path: &Option<String>,
+    );
+
+    fn handle_responses(
+        &mut self, conn: &mut quiche::Connection, buf: &mut [u8],
+        req_start: &std::time::Instant,
+    );
+
+    fn report_incomplete(&self, start: &std::time::Instant);
+
+    fn handle_requests(
+        &mut self, conn: &mut std::pin::Pin<Box<quiche::Connection>>,
+        partial_responses: &mut HashMap<u64, PartialResponse>, root: &str,
+        buf: &mut [u8],
+    ) -> i8;
```

Comment:
This is not C :) so it should return `Result`. It seems the only error condition is when quiche has an error, so you can probably just use `quiche::Result<()>` and simply propagate the quiche error on failure, and return `Ok(())` on success.

Reply:
ha yeah this is my fault. I did try to use result to start with but it really screwed something up (the loop breaking maybe?). Now the code has been refactored then it might be easy to put this in.

Reply:
In practice you can basically do the same thing you are doing now, except that instead of checking ` == -1` you can just do `.is_err()`. e.g.
```rust
if blah().is_err() {
  // handle error
}
```